// @elysiajs/jwt/dist/index.mjs
// src/index.ts
import { Elysia, t, ValidationError, getSchemaValidator } from "elysia";
import {
  SignJWT,
  jwtVerify
} from "jose";
var jwt = ({
  name = "jwt",
  secret,
  schema,
  ...defaultValues
}) => {
  if (!secret) throw new Error("Secret can't be empty");
  const key = typeof secret === "string" ? new TextEncoder().encode(secret) : secret;
  const validator = schema ? getSchemaValidator(schema, {
    modules: t.Module({}),
    validators: [
      t.Object({
        iss: t.Optional(t.String()),
        sub: t.Optional(t.String()),
        aud: t.Optional(
          t.Union([t.String(), t.Array(t.String())])
        ),
        jti: t.Optional(t.String()),
        nbf: t.Optional(t.Number()),
        exp: t.Optional(t.Number()),
        iat: t.Optional(t.Number())
      })
    ]
  }) : void 0;
  return new Elysia({
    name: "@elysiajs/jwt",
    seed: {
      name,
      secret,
      schema,
      ...defaultValues
    }
  }).decorate(name, {
    sign(signValue) {
      const { nbf, exp, iat, ...data } = signValue;
      const JWTHeader = {
        alg: defaultValues.alg ?? "HS256",
        b64: defaultValues.b64,
        crit: defaultValues.crit,
        cty: defaultValues.cty,
        jku: defaultValues.jku,
        jwk: defaultValues.jwk,
        kid: defaultValues.kid,
        typ: defaultValues.typ ?? "JWT",
        x5c: defaultValues.x5c,
        x5t: defaultValues.x5t,
        x5u: defaultValues.x5u
      };
      const JWTPayload = {
        /**
         * Audience (aud): Identifies the recipients that the JWT is intended for.
         */
        aud: data.aud ?? defaultValues.aud,
        /**
         * Issuer (iss): Identifies the principal that issued the JWT.
         */
        iss: data.iss ?? defaultValues.iss,
        /**
         * JWT ID (jti): Provides a unique identifier for the JWT.
         */
        jti: data.jti ?? defaultValues.jti,
        /**
         * Subject (sub): Identifies the principal that is the subject of the JWT.
         */
        sub: data.sub ?? defaultValues.sub,
        // Includes all other properties from the data source, both standard and custom,
        // excluding standard JWT claims like `nbf`, `exp` and `iat`.
        ...data
      };
      let jwt2 = new SignJWT({ ...JWTPayload }).setProtectedHeader({
        alg: JWTHeader.alg,
        ...JWTHeader
      });
      const setNbf = "nbf" in signValue ? nbf : defaultValues.nbf;
      if (setNbf !== void 0) jwt2 = jwt2.setNotBefore(setNbf);
      const setExp = "exp" in signValue ? exp : defaultValues.exp;
      if (setExp !== void 0) jwt2 = jwt2.setExpirationTime(setExp);
      const setIat = "iat" in signValue ? iat : defaultValues.iat;
      if (setIat !== false) jwt2 = jwt2.setIssuedAt(/* @__PURE__ */ new Date());
      return jwt2.sign(key);
    },
    async verify(jwt2, options) {
      if (!jwt2) return false;
      try {
        const data = (await (options ? jwtVerify(jwt2, key, options) : jwtVerify(jwt2, key))).payload;
        if (validator && !validator.Check(data))
          throw new ValidationError("JWT", validator, data);
        return data;
      } catch {
        return false;
      }
    }
  });
};
var index_default = jwt;
export {
  index_default as default,
  jwt
};
