// @elysiajs/jwt/dist/cjs/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  default: () => index_default,
  jwt: () => jwt
});
module.exports = __toCommonJS(index_exports);
var import_elysia = require("elysia");
var import_jose = require("jose");
var jwt = ({
  name = "jwt",
  secret,
  schema,
  ...defaultValues
}) => {
  if (!secret) throw new Error("Secret can't be empty");
  const key = typeof secret === "string" ? new TextEncoder().encode(secret) : secret;
  const validator = schema ? (0, import_elysia.getSchemaValidator)(schema, {
    modules: import_elysia.t.Module({}),
    validators: [
      import_elysia.t.Object({
        iss: import_elysia.t.Optional(import_elysia.t.String()),
        sub: import_elysia.t.Optional(import_elysia.t.String()),
        aud: import_elysia.t.Optional(
          import_elysia.t.Union([import_elysia.t.String(), import_elysia.t.Array(import_elysia.t.String())])
        ),
        jti: import_elysia.t.Optional(import_elysia.t.String()),
        nbf: import_elysia.t.Optional(import_elysia.t.Number()),
        exp: import_elysia.t.Optional(import_elysia.t.Number()),
        iat: import_elysia.t.Optional(import_elysia.t.Number())
      })
    ]
  }) : void 0;
  return new import_elysia.Elysia({
    name: "@elysiajs/jwt",
    seed: {
      name,
      secret,
      schema,
      ...defaultValues
    }
  }).decorate(name, {
    sign(signValue) {
      const { nbf, exp, iat, ...data } = signValue;
      const JWTHeader = {
        alg: defaultValues.alg ?? "HS256",
        b64: defaultValues.b64,
        crit: defaultValues.crit,
        cty: defaultValues.cty,
        jku: defaultValues.jku,
        jwk: defaultValues.jwk,
        kid: defaultValues.kid,
        typ: defaultValues.typ ?? "JWT",
        x5c: defaultValues.x5c,
        x5t: defaultValues.x5t,
        x5u: defaultValues.x5u
      };
      const JWTPayload = {
        /**
         * Audience (aud): Identifies the recipients that the JWT is intended for.
         */
        aud: data.aud ?? defaultValues.aud,
        /**
         * Issuer (iss): Identifies the principal that issued the JWT.
         */
        iss: data.iss ?? defaultValues.iss,
        /**
         * JWT ID (jti): Provides a unique identifier for the JWT.
         */
        jti: data.jti ?? defaultValues.jti,
        /**
         * Subject (sub): Identifies the principal that is the subject of the JWT.
         */
        sub: data.sub ?? defaultValues.sub,
        // Includes all other properties from the data source, both standard and custom,
        // excluding standard JWT claims like `nbf`, `exp` and `iat`.
        ...data
      };
      let jwt2 = new import_jose.SignJWT({ ...JWTPayload }).setProtectedHeader({
        alg: JWTHeader.alg,
        ...JWTHeader
      });
      const setNbf = "nbf" in signValue ? nbf : defaultValues.nbf;
      if (setNbf !== void 0) jwt2 = jwt2.setNotBefore(setNbf);
      const setExp = "exp" in signValue ? exp : defaultValues.exp;
      if (setExp !== void 0) jwt2 = jwt2.setExpirationTime(setExp);
      const setIat = "iat" in signValue ? iat : defaultValues.iat;
      if (setIat !== false) jwt2 = jwt2.setIssuedAt(/* @__PURE__ */ new Date());
      return jwt2.sign(key);
    },
    async verify(jwt2, options) {
      if (!jwt2) return false;
      try {
        const data = (await (options ? (0, import_jose.jwtVerify)(jwt2, key, options) : (0, import_jose.jwtVerify)(jwt2, key))).payload;
        if (validator && !validator.Check(data))
          throw new import_elysia.ValidationError("JWT", validator, data);
        return data;
      } catch {
        return false;
      }
    }
  });
};
var index_default = jwt;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  jwt
});
